### DISCUSSION TITLE
basic-types/char-bool-unit

### DISCUSSION THREAD

**Discussion by giscus** [sunface/rust-by-practice#182] [Book Comments]
# basic-types/char-bool-unit

Learning Rust By Practice, narrowing the gap between beginner and skilled-dev with challenging examples, exercises and projects.

https://zh.practice.rs/basic-types/char-bool-unit.html

**Comment by mimizao** [comment]
不知道第五行是否少了一个分号，我猜想这应该不是您想考察的地方
```rust
fn main() {
    let _v: () = ();

    let v = (2, 3);
    assert_eq!(v, implicitly_ret_unit()) 
    println!("Success!")
}
```

**Reply by sunface** [reply]
是的，少了，应该是因为最后一行是后面复制过去的

**Reply by roushanyyzz** [reply]
return ()，所以不需要分号

**Reply by zhanghel1ng** [reply]
只有最后一行不需要分号

**Reply by c3qu** [reply]
加上分号返回(),不加分号print!返回也是().加不加分号效果都一样

**Reply by Zhaoyikaiii** [reply]
在 Rust 中，语句（Statements）和表达式（Expressions）有一些关键的区别。

1. **表达式（Expressions）**：表达式是计算出某个值的代码片段。例如，`5 + 6` 是一个表达式，因为它计算出值 `11`。函数调用（如 `println!("Hello, world!")`）和宏调用也是表达式。块 `{}`，如果它们以表达式结尾，也是表达式。
2. **语句（Statements）**：语句是执行某些操作但不返回值的代码片段。例如，`let x = 5;` 是一个语句，因为它创建了一个变量 `x`，但没有返回值。

在 Rust 中，函数体是由一系列语句构成的。如果你想让函数返回一个值，你需要在函数体的最后放一个表达式，而不是语句。这个表达式的值就会成为函数的返回值。注意，这个表达式后面不能有分号，因为在 Rust 中，分号会把表达式转变为语句，而语句不返回值。

例如，以下函数返回一个 `i32` 类型的值：

```rust
fn add_two(x: i32) -> i32 {
	x + 2  // 这是一个表达式，它的值会被返回
}
```

如果你在这个表达式后面加上分号，它就变成了一个语句，函数就不会返回值了：

```rust
fn add_two(x: i32) -> i32 {
  x + 2;  // 这是一个语句，它不返回值
}
```

这段代码会导致编译错误，因为函数声明中指定了返回类型 `i32`，但函数体中没有返回值。

**Comment by fireindark707** [comment]
完成 6/6

**Reply by RomanSeven7** [reply]
done  ! ! !

**Comment by mikucy** [comment]
完成！第五题的I will returen...是不是手抖拼错了return😳

**Comment by lovefengruoqing** [comment]
checked!

**Comment by sdllh** [comment]
请问单元类型不占用内存，那么他存放在哪里呢？

**Reply by NicholasFirst** [reply]
这里元组是空的所以不占内存, 你试着加几个元素进去再试试看?

**Reply by Tmem0rethinks** [reply]
只是声明，还没有分配内存，声明应该是不放在内存中，因为内存中一般是为了快速运行。目前它是放在你这个文件中，这是我的理解。

**Comment by guanguoliang** [comment]
小白问一下，第一题size_of_val的结果为啥是4，'a'一个字符不是只有一个字节吗

**Reply by guanguoliang** [reply]
重温了知识，字符类型就是4个字节，代表Unicode标量值

**Reply by ZYKWLJ** [reply]
没错，字面量是每个4byte，但是字符串不是，1-4任选

**Comment by Elixin** [comment]
我需要戴眼镜了🧐，第二题我看半天，还以为char有特殊的输出方式，结果。。。

**Reply by xiaoyue2019** [reply]
我也看了半天答案...

**Reply by ytxcash** [reply]
-> 注意，我们还没开始讲字符串，但是这里提前说一下，和一些语言不同，Rust 的字符只能用 '' 来表示， "" 是留给字符串的

我是回去翻的教程

**Reply by ZYKWLJ** [reply]
这样改也可以：
```rust
// 修改一行让代码正常打印
fn main() {
    let c1 = '中';
    print_char(c1);
} 

fn print_char(c : &str) {
    println!("{}", c);
}


```

**Comment by makexiao113** [comment]
done！

**Comment by jackyliu16** [comment]
想问下第一题的地方，是不是这个size_of_val()
是因为**传入的是一个指针**然后这个指针的大小是4B，所以说返回了这个结果？

**Reply by jackyliu16** [reply]
我是傻逼

**Reply by xiaoyue2019** [reply]
哈哈哈 🤣

**Comment by makexiao113** [comment]
我的理解是汉字是两个byte，一个byte是4bit，两个汉字共16bit



---Original---
From: ***@***.***&gt;
Date: Wed, Jul 6, 2022 00:08 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [sunface/rust-by-practice] basic-types/char-bool-unit (Discussion#182)




 
想问下第一题的地方，是不是这个size_of_val()
 是因为传入的是一个指针然后这个指针的大小是4B，所以说返回了这个结果？

**Reply by Unik-lif** [reply]
一般情况下确实256*256两个字节来表示汉字，但是Rust用的是Unicode编码，默认每个字符都是4字节。

**Comment by makexiao113** [comment]
额，我应该搞错了，rust字符保存的是unicode，一个汉字是2个unicode共8字节
，2个汉字共16字节。我也不是很确定

---Original---
From: ***@***.***&gt;
Date: Wed, Jul 6, 2022 00:11 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [sunface/rust-by-practice] basic-types/char-bool-unit (Discussion #182)




 
我是傻逼

**Comment by AnranS** [comment]
```
fn main() {
    let c1 = '中';  // "" => '' 麻了
    print_char(c1);
} 

fn print_char(c : char) {
    println!("{}", c);
}
```

**Comment by zonghenghaunyu** [comment]
lm

**Comment by fatFire** [comment]
第5题 implicitly_ret_unit() 返回空元祖吗？ ()

**Reply by cxw620** [reply]
原文如下哦
```
例如常见的 println!() 的返回值也是单元类型 ()。

**Comment by TheBigFish** [comment]
done

**Comment by Grapeve** [comment]
2022.11.9 打卡

**Comment by twandylue** [comment]
Done

**Comment by jxderic** [comment]
done

**Comment by Bruce012356** [comment]
mark

**Comment by Jeraks** [comment]
第五题这样也行？？？这到底比了个啥？？是不是可以拆开写🤔

```
fn main() {
    let _v: () = explicitly_ret_unit();

    let v = (2, 3);
    assert_eq!(_v, implicitly_ret_unit());

    println!("Success!")
}

fn implicitly_ret_unit() {
    println!("I will return a ()")
}

// 不要使用下面的函数，它只用于演示！
fn explicitly_ret_unit() -> () {
    println!("I will return a ()")
}

```

**Reply by samwu97** [reply]
`implicitly_ret_unit` 没有写返回值，默认返回空元组`()`。

**Reply by AlexanderChou9** [reply]
也可以给第一个 v 去掉下划线，给第二个 v 加上下划线，它会自动用前面那个。编译器还挺智能

**Comment by Kori000** [comment]
done 4/5

**Comment by cddsgtc** [comment]
pass

**Comment by rain0073** [comment]
6/6,all done!

**Comment by hexWars** [comment]
第五题答案有误，题目是`_v`答案是`v0`

**Reply by RenXusheng233** [reply]
这里应该是故意改的，因为 `_v` 意味着告诉 Rust 这个是未使用的变量。

**Comment by XB-Echo** [comment]
6/6

**Comment by XU0618** [comment]
lala~（好喜欢打印的Success！

**Comment by 1576584678** [comment]
全部 Success!

**Comment by owocc** [comment]
完成!!

**Comment by zhyuxi** [comment]
个人理解第二题的另一种解法
```
// 修改一行让代码正常打印
fn main() {
    let c1 = "中";
    print_char(c1);
} 

fn print_char(c : &str) {
    println!("{}", c);
}
```

**Reply by killdayu** [reply]
print_char

**Comment by XavierZXY** [comment]
checked!

**Comment by aisyser** [comment]
# done!!!

**Comment by Hangghost** [comment]
第二題的答案沒有改正？

**Reply by Hangghost** [reply]
我是改成：
c : &str

**Reply by Hangghost** [reply]
抱歉，沒事，我矇了！

**Comment by chenhai998** [comment]
done

**Comment by 4may-mcx** [comment]
done~

**Comment by Ray-lyy** [comment]
done，谢谢作者大大的无私奉献！

**Comment by CynicalHeart** [comment]
mark

**Comment by floiges** [comment]
完美

**Comment by Muzych** [comment]
MARK，目前存在的疑问就是单元类型到底是个啥，它有啥用

**Comment by Day333** [comment]
Done!

**Comment by Ricardo-LLH** [comment]
我就是只有一点点编程基础来学rust的好汉，确实有点头皮发麻：（，我看到最后一题的答案v0，我以为是rust中什么特殊的语法，后来才知道这是让我在单元类型和元组选一个。我是傻逼

**Comment by xiang-noodle** [comment]
done--2023/12/9

**Comment by thiru-01** [comment]
done

**Comment by TomContral** [comment]
Done

**Comment by vickyone0** [comment]
Done batman

**Comment by vickyone0** [comment]
done batman

**Comment by zhengkezhou1** [comment]
done

**Comment by EmptyBlueBox** [comment]
checked!

**Comment by TDAkory** [comment]
mark finished

**Comment by sovansela** [comment]
--completed

**Comment by xiryg** [comment]
**done**

**Comment by Prajwalraut29** [comment]
done

**Comment by kid1110** [comment]
done

**Comment by DylanJinx** [comment]
done~~~

**Comment by C-RONALDO7V** [comment]
accepted

**Comment by sjunepark** [comment]
Done

**Comment by H-Felicity** [comment]
字符类型也是占用 4 个字节、布尔值占用内存的大小为 1 个字节、单元类型完全不占用任何内存

**Comment by H-Felicity** [comment]
Rust 的字符只能用 '' 来表示， "" 是留给字符串的。
字符类型也是占用 4 个字节、布尔值占用内存的大小为 1 个字节、单元类型完全不占用任何内存。

**Comment by Murasame-BH8HJY** [comment]
Done!

**Comment by lhylyrr** [comment]
DONE!

**Comment by lichenbuliren** [comment]
done

**Comment by Sagecheni** [comment]
DONE!

**Comment by hash-panda** [comment]
println!("Done");

**Comment by zhengkezhou1** [comment]
day1

**Comment by Paul-from-NJTech** [comment]
Done!

**Comment by MahajanSamiksha** [comment]
done

**Comment by UnclePhil1** [comment]
I'm done

**Comment by bansal-harsh-2504** [comment]
✅

**Comment by gastrodia** [comment]
Success(6/6)

**Comment by chttiu** [comment]
标记完成

**Comment by mdShakilHossainNsu2018** [comment]
done

**Comment by eyenov** [comment]
done
